feat: junit cli test report format

mmadson opened the pull request:
WIP: mainly to get the ball rolling and start the discussion. 

Schema for Junit XML is at:

- https://llg.cubic.org/docs/junit/

Seems that the junit xml report has quite a few fields that our json output doesn't currently support. Not sure how difficult it would be to capture. Then again not all of the junit xml fields are required, so we can probably get pretty far by starting with only the required fields. 

Also I'm sure the code organization here is bad. The util function should not depend upon the the core domain class of the SendJsonOutput type but figured we can iterate and figure out where stuff should go in this draft MR.

AnWeber commented on src/utils/junitUtils.ts:
```diff
@@ -0,0 +1,40 @@
+`;
+    for (const testResult of req.testResults ?? []) {
+      resultXml += `
+      <testcase 
```
I am missing a level here. For me, the test file (*.http) but also each individual region would be in this file. Then you could output each test result as <error> or <failure>.

AnWeber replied:
In your example on Google Drive, you can actually see what bothers me about it. The properties give the name of the current region. The test case below the properties belongs to this region. I think it would be better encapsulated in a testsuite. How do other reporters do it for jest for example?

mmadson replied:
Sorry I'm not following the concern. I think all the information is present. The testsuite package attribute is used for the file path. The testsuite name is used for the region and the child elements of the testsuite capture the test case names / conditions / assertions. 

What level are we missing here?

mmadson replied:
to answer your question, it looks like jest adopts quite an extensive list of configuration options to help map between jest test conventions and junit xml with as sensible a set of defaults as possible. I think it would be good to try and support some additional configuration as well for this feature -- though perhaps, that can be a follow on merge request.

https://github.com/palmerj3/jest-junit#configuration

AnWeber replied:
The properties in your video were at the `<testsuites>` level and therefore also before the `<testsuite>` they belong to. That's what I would have meant. Now I pulled them into testcase and I think they are correct there, even if it is not quite compliant with the XML.

> support some additional configuration as well for this feature -- though perhaps, that can be a follow on merge request.

Yes, let's integrate the feature and then see what else is needed when used. I still spontaneously missing stdout log and possibly the response, but that may simply clutter up the file.

AnWeber commented on src/utils/junitUtils.ts:
```diff
@@ -0,0 +1,40 @@
+      name="${req.fileName}"
+      tests="${req.summary.totalTests}"
+      disabled="0"
+      errors="${req.testResults?.filter(r => r.error !== undefined).length}"
```
I check for r.result, https://github.com/AnWeber/httpyac/blob/main/src/cli/send/jsonOutput.ts#L63
```suggestion
      errors="${req.testResults?.filter(r => °r.result).length}"
```

AnWeber commented:
fantastic work. Thx for your input.

mmadson commented:
I believe required tags / attributes and core report logic is now in place. I think what remains is cleanup, refactoring and tests. Here is a demo of the feature in action comparing between json and junit xml output.

- https://drive.google.com/file/d/17HraIiyccgO3RVaIQVxm5ecmS15_Jype/view?usp=sharing

mmadson commented on src/utils/junitUtils.ts:
```diff
@@ -0,0 +1,68 @@
+        resultXml += `</failure>`;
+        resultXml += `</testcase>`;
+      } else {
+        resultXml += '/>';
```
autoclosed single line testcase if there is no failure info to report

mmadson commented:
- added execute interceptors to capture the duration of the request using the js performance api.
- added to json output format
- converted to float seconds for the junit xml report

AnWeber commented on src/utils/junitUtils.ts:
```diff
@@ -0,0 +1,68 @@
+}
+
+export function toJunitXml(output: SendJsonOutput): string {
+  let resultXml = `<?xml version="1.0" encoding="UTF-8"?><testsuites errors="0" failures="${output.summary.failedTests}" tests="${output.summary.totalTests}">`;
```
Hi yes I'm thinking of checking if it's an [AssertionError](https://nodejs.org/api/assert.html#new-assertassertionerroroptions) and only considering these as test errors. but that can also mean false positives, that's why I threw them together currently.

mmadson replied:
Sounds good, I don't think it's a deal breaker to defer this for now but your call.

AnWeber replied:
Yes, I think so too. I also fear that the check is not quite accurate, as you are tempted to simply generate an error and not expect anything of a special logic per error type.

AnWeber commented on src/models/httpResponse.ts:
```diff
@@ -18,6 +18,7 @@ export interface HttpResponse {
   request?: Request;
 
   timings?: HttpTimings;
+  durationMillis?: number;
```
The Property seems out of place here. The duration does not show the duration of the response. But rather the duration for the whole processing of the httpRegion. 
My idea would be to move the creation of the data object (ProcessedHttpRegion) to a single interceptor and therefore avoid these properties.

mmadson replied:
I see, so we want to move the logic of creating toProcessedHttpRegion out of logResponseInterceptor into a new processedHttpRegionInterceptor and have the beforeLoop set the start duration and the afterLoop take the completed httpResponse and generate the final ProcessedHttpRegion is that correct?

AnWeber replied:
yes, this was my idea. One small Problem. The logResponseInterceptor should create a copy of the response and call logResponse Hook. This copy is also needed in the new processedHttpRegionInterceptor. We need a nice way to geht this response.

AnWeber commented on src/plugins/core/execute/index.ts:
```diff
@@ -4,11 +4,15 @@ import { DisabledInterceptor } from './disabledInterceptor';
 import { LazyVariableInterceptor } from './lazyVariableInterceptor';
 import { LogResponseInterceptor } from './logResponseInterceptor';
 import { RegionScopedVariablesInterceptor } from './regionScopedVariablesInterceptor';
+import { RequestDurationEndInterceptor } from './requestDurationEndInterceptor';
```
As noted, an Interceptor would probably be better. However, I am not sure if this will then become the LogResponseInterceptor. This would have in this case already two tasks (trigger logResponse hook, manage ProcessedHttpRegion) and that seems to me also not right.

mmadson replied:
I think I'm a bit confused on the recommendation here. Would you like me to move the duration logic to the `LogResponseInterceptor` adding a beforeLoop to the class to start the performance recording?

If so, I suppose my question is where should I store the start of the performance timing duration in the beforeLoop? I'm currently saving it on the response object but agree that's probably not the best spot for it. Should it be a new top level context variable -- maybe something like context.timings or context.metrics? Or did you want me to initialize the ProcessedHttpRegion in the before, save off the duration start and then modify the processed http region in the afterloop?

Appreciate the clarification.

AnWeber replied:
In the context is a options object which can store the metadata, maybe de should use this object. I will contribute to your branch to test some ideas.

AnWeber replied:
Using the context.options object would have disadvantages that I overlooked. When using `# @ref` the context is cloned and the options object is passed by reference. But this would falsify the duration stored in the options object. For this reason I now create the ProcessedHttpRegion in the beforeLoop and load it again from the array using the id of the current HttpRegion.

AnWeber commented on src/utils/junitUtils.ts:
```diff
@@ -0,0 +1,76 @@
+import { SendJsonOutput } from '@/cli/send/jsonOutput';
+
+const xmlSpecialCharReplacementMap: Record<string, string> = {
```
I already have [xmldom reference](https://github.com/AnWeber/httpyac/blob/main/package.json#L99) in the project. I am considering building this string concatenation story, with the framework and then also [outputting it using with `formatXML`](https://github.com/AnWeber/httpyac/blob/main/src/plugins/xml/xmlResponseInterceptor.ts#L3).
I would compare it once to see if it reads better, but it would probably solve the escaping issues directly.

AnWeber replied:
To me xmldom looks nicer. Then you don't have to worry about XML output at the same time.

AnWeber commented on src/utils/junitUtils.ts:
```diff
@@ -0,0 +1,76 @@
+import { SendJsonOutput } from '@/cli/send/jsonOutput';
```
The import should be changed. I am afraid of an import loop with this. I would also definitely drag junitUtils over to cli, since this is only needed there.

AnWeber commented on package.json:
```diff
@@ -110,7 +110,7 @@
     "got": "^11.8.6",
     "grpc-reflection-js": "^0.3.0",
-    "hookpoint": "3.2.1",
+    "hookpoint": "4.0.0",
```
I have built an update for Hookpoint. Now you can see in the Interceptor in the `afterLoop` if the region is disabled ( if hookContext.bail is truthy), because this method is still called then. Say I would still look there to extend this.

mmadson commented on src/cli/send/jsonOutput.ts:
```diff
@@ -38,52 +43,57 @@ function sum(x: number, y: number) {
+export function createTestSummary(requests: Array<SendOutputRequest>): SendRequestSummary & SendTestSummary {
+  return {
+    totalRequests: requests.length,
+    disabledRequests: requests.filter(obj => obj.disabled).length,
```
awesome! 😮

mmadson commented on src/cli/send/junitUtils.spec.ts:
```diff
@@ -0,0 +1,216 @@
+      `
+<testsuites errors="0" disabled="0" failues="0">
+  <testsuite name="test.http" tests="2" failures="0" skipped="0" package="." time="2.003" file="test.http">
+    <properties>
```
the properties node here is a bit strange. I would expect that a testsuite has at most 1 properties node but the properties node can have multiple property child nodes. It seems like we are duplicating the node without nesting into a test case to try and imply association based on position of properties node in relation to the test case in the output,

AnWeber replied:
testcase unfortunately has no properties element. I would now use the same hack as [playwright](https://github.com/microsoft/playwright/blob/c3b533d8341d72d45b7296c7a895ff9fe7d8ff3b/packages/playwright/src/reporters/junit.ts#L150-L170) and just put the properties below testcase.

mmadson commented on src/cli/send/junitUtils.spec.ts:
```diff
@@ -0,0 +1,216 @@
+    });
+    expect(result).toBe(
+      `
+<testsuites errors="0" disabled="0" failues="0">
```
if possible it would be good to include the summary information in the testsuites node attributes -- tests being the total number of tests across all child testsuite nodes. time to indicate total time it took to run all tests in all child test suites. Also having a name for the overall testsuites collection like (httpyac tests) would be useful for reporting.

AnWeber replied:
attributes added

mmadson commented on src/cli/send/junitUtils.spec.ts:
```diff
@@ -0,0 +1,216 @@
+    });
+    expect(result).toBe(
+      `
+<testsuites errors="0" disabled="0" failues="1">
```
when we write to stdout will the xml frontmatter be included or do we need to do something to add:

`<?xml version="1.0" encoding="UTF-8"?>`

AnWeber replied:
xml formatter added using `createProcessingInstruction`

mmadson commented on src/plugins/core/execute/processedHttpRegionInterceptor.ts:
```diff
@@ -0,0 +1,69 @@
+  ): Promise<boolean | undefined> {
+    const [context] = hookContext.args;
+
+    const processedHttpRegion = context.processedHttpRegions?.find(obj => obj.id === context.httpRegion.id);
```
wonderful, was wondering how you were going to handle this. New ID field is very useful.

AnWeber replied:
However, I find the ID dangerous. It depends on the line in the file and thus is not always unique if the file changes. This is not a problem for this case. but I am slightly afraid that I forget the restriction.

mmadson commented on src/plugins/core/execute/processedHttpRegionInterceptor.ts:
```diff
@@ -0,0 +1,69 @@
+      filename: context.httpFile.fileName,
+      symbol: context.httpRegion.symbol,
+      isGlobal: context.httpRegion.isGlobal(),
+      start: new Date().getTime(),
```
purely for my own education, why opt for date time vs performance api? It looks like the performance api is more granular at the cost of being slower:

https://www.sitepoint.com/discovering-the-high-resolution-time-api/

AnWeber replied:
I actually wanted to change that back. Yes, I agree with you. `performance.now()` would be better, but I also wanted to output the current time and performance.now doesn't provide that. Ic didn't know the API and quickly changed it to Date. Now it is correct.

AnWeber commented:
@mmadson I think we have covered all the necessary cases for the first draft. Do you still notice any problems? Otherwise, I would integrate it like this.

mmadson commented:
This looks exceptional. Thank you so much for not only being willing to accommodate this feature but being so active in developing it. Truly an remarkable OSS maintainer; thank you! I assume if we overlooked anything with the report generation in terms of integrating the output with gitlab, we can patch it up fairly quick with a follow on MR. So, everything looks good on my end. Excited to give it a try in my test project's functional test pipeline!

AnWeber commented:
I am happy to return the praise and thanks. Thank you for taking the time to read and improve my code. And of course we can react on short notice in case of problems.
